### PR TITLE
[clang][DFP] Add support for -fexperimental-decimal-floating-point.

### PR DESCRIPTION
**Pull request by tahonermann**
Though the ISO/IEC TS 18661-2:2015 decimal floating-point (DFP) extension has been adopted for the upcoming C23 standard, it will remain necessary to be able to explicitly enable and disable support for the extension for the foreseeable future for use with previous C standards and with C++. This change adds a driver and cc1 option to enable language support. Since support for the feature is not yet complete, the option name explicitly reflects its status as experimental and only serves to enable testing at this point.

No target considerations have been implemented yet. The option only serves to enable the _Decimal32, _Decimal64, and _Decimal128 keywords to be used as DFP type specifiers in C; these keywords are not enabled for C++ since, per ISO/IEC TR 24733:2011 and the gcc implementation, DFP support in C++ is provided via standard library types (that are expected to wrap a builtin type exposed by other means).

### CODE REVIEW

File: clang/lib/Frontend/InitPreprocessor.cpp [hunk 1]
```diff
@@ -498,6 +498,12 @@ static void InitializeStandardPredefinedMacros(const TargetInfo &TI,
   Builder.defineMacro("__STDC_UTF_16__", "1");
   Builder.defineMacro("__STDC_UTF_32__", "1");
 
+  // C23 decimal floating point extensions.
+  // FIXME: Define to 202311L when support for C23 decimal floating point
+  // FIXME: extensions is feature complete.
+  if (!LangOpts.CPlusPlus && LangOpts.DecimalFloatingPoint)
+    Builder.defineMacro("__STDC_IEC_60559_DFP__", "197001L");
```
Comment (tahonermann) [inline clang/lib/Frontend/InitPreprocessor.cpp:505]: Issue #9 tracks support for predefined macros. The value defined here is a placeholder but will suffice for us to use the macro for test purposes.

File: clang/lib/Sema/SemaType.cpp [hunk 2]
```diff
@@ -1588,9 +1588,13 @@ static QualType ConvertDeclSpecToType(TypeProcessingState &state) {
...
-    S.Diag(DS.getTypeSpecTypeLoc(), diag::err_decimal_unsupported);
-    Result = TSTToDecimalFloatType(Context, DS.getTypeSpecType());
-    declarator.setInvalidType(true);
+    if (!S.getLangOpts().DecimalFloatingPoint) {
+      S.Diag(DS.getTypeSpecTypeLoc(), diag::err_dfp_disabled);
+      Result = Context.IntTy;
+      declarator.setInvalidType(true);
+    } else {
+      Result = TSTToDecimalFloatType(Context, DS.getTypeSpecType());
+    }
```
Comment (tahonermann) [inline clang/lib/Sema/SemaType.cpp:1597]: I have other changes that will be submitted in another PR that will add checking for target support and a diagnostic for attempted DFP use on a target that lacks support.

File: clang/test/Driver/dfp-enablement-lang.c [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// This test is intended to validate whether decimal floating-point (DFP)
+// extensions are enabled based on language standard and to ensure a proper
+// diagnostic is issued for any use of DFP features otherwise.
+
+// RUN: %clang -target x86_64-unknown-linux-gnu -std=c17 -fsyntax-only -Xclang -verify=dfp-off,c-dfp-off %s
+// RUN: %clang -target x86_64-unknown-linux-gnu -std=c17 -fsyntax-only -fexperimental-decimal-floating-point -Xclang -verify %s
+// FIXME: Remove -fexperimental-decimal-floating-point once -std=c23 implies DFP enablement.
```
Comment (tahonermann) [inline clang/test/Driver/dfp-enablement-lang.c:7]: Issue #22 tracks making the change to `-std=c23` and removing what will then be extraneous uses of `-fexperimental-decimal-floating-point`.

File: clang/test/Driver/dfp-enablement-lang.c [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// RUN: %clang -target x86_64-unknown-linux-gnu -std=c23 -fsyntax-only -fexperimental-decimal-floating-point -Xclang -verify %s
+// RUN: %clang -target x86_64-unknown-linux-gnu -std=c23 -fsyntax-only -fno-experimental-decimal-floating-point -Xclang -verify=dfp-off,c-dfp-off %s
+// RUN: %clang -target x86_64-unknown-linux-gnu -x c++ -std=c++23 -fsyntax-only -Xclang -verify=cxx,dfp-off %s
+// RUN: %clang -target x86_64-unknown-linux-gnu -x c++ -std=c++23 -fsyntax-only -fexperimental-decimal-floating-point -Xclang -verify=cxx %s
+// FIXME: Remove -fexperimental-decimal-floating-point once -std=c++2c (or later) implies DFP enablement.
```
Comment (tahonermann) [inline clang/test/Driver/dfp-enablement-lang.c:12]: As far as I know, there is no current proposal to add DFP support to the C++ standard.

**Review by zahiraam — APPROVED** [review]
Looks good. Thanks.

**Review by shafik — APPROVED** [review]
LGTM

**Comment by tahonermann** [comment]
This ended up landing as commit https://github.com/tahonermann/llvm-project/commit/f06fce6d01508456a91dc702d4a92e313bc06f47 because I'm still confused by GitHub's squash and merge interface that populates the commit message in a way that ends up not doing what I expect; I amended the commit message after the changes were merged.
